yb-voyager/src/queryparser/traversal_plpgsql.go

```diff
@@ -228,3 +216,183 @@ func formatExprQuery(q string) string {
+                        }
+                    }
+                },
+                {
+                    "PLpgSQL_var": {
+                        "refname": "found",
+                        "datatype": {
+                            "PLpgSQL_type": {
+                                "typname": "UNKNOWN"
+                            }
+                        }
+                    }
+                },
+                {
+                    "PLpgSQL_var": {
+                        "refname": "emp_salary",
+                        "lineno": 3,
+                        "datatype": {
+                            "PLpgSQL_type": {
+                                "typname": "employees.salary%TYPE"
+                            }
+                        }
+                    }
+                },
+                {
+                    "PLpgSQL_row": {
+                        "refname": "(unnamed row)",
+                        "lineno": 5,
+                        "fields": [
+                            {
+                                "name": "emp_salary",
+                                "varno": 2
+                            }
+                        ]
+                    }
+                }
+            ],"action": {
+               ....
+            }
+        }
+    },
+
+	Caveats:
+	1. Not returning typename for variables in function parameter from this function (in correct in json as UNKNOWN), for that using the GetTypeNamesFromFuncParameters()
+	2. Not returning the return type from this function (not available in json), for that using the GetReturnTypeOfFunc()
+*/
+func GetAllTypeNamesInPlpgSQLStmt(query string) ([]string, error) {
+	parsedJson, parsedJsonMap, err := getParsedJsonMap(query)
+	if err != nil {
+		return []string{}, nil
+	}
+	function := parsedJsonMap[PLPGSQL_FUNCTION]
+	parsedFunctionMap, ok := function.(map[string]interface{})
+	if !ok {
+		return []string{}, fmt.Errorf("error getting the PlPgSQL_Function field in parsed json-%s", parsedJson)
+	}
+
+	datums := parsedFunctionMap[DATUMS]
+	datumList, isList := datums.([]interface{})
+	if !isList {
+		return []string{}, fmt.Errorf("error getting type names datums field is not list in parsed json-%s", parsedJson)
+	}
+
+	var typeNames []string
+	for _, datum := range datumList {
+		datumMap, ok := datum.(map[string]interface{})
+		if !ok {
+			continue
+		}
+		for key, val := range datumMap {
+			switch key {
+			case PLPGSQL_VAR:
+				typeName := getTypeNameFromPlpgSQLVar(val)
+				if typeName != "" {
+					typeNames = append(typeNames, typeName)
+				}
+			}
+		}
+	}
+	return typeNames, nil
+}
+
+/*
+example of PLPGSQL_VAR -
+"PLpgSQL_var": {
+	"refname": "tax_rate",
+	"lineno": 3,
+	"datatype": {
+		"PLpgSQL_type": {
+			"typname": "employees.tax_rate%TYPE"
+		}
+	}
+}
+*/
+func getTypeNameFromPlpgSQLVar(plpgsqlVar interface{}) string {
+	//getting the map of <key,val > of PLpgSQL_Var json
+	valueMap, ok := plpgsqlVar.(map[string]interface{})
+	if !ok {
```

Comment:
In this function, for all the `!ok`, how about we just return an error. The function should either successfully get the type name from the plpgSQLVar or return error. Why return empty string and expect the caller to check if it's empty or not?

In this case, I guess the caller wouldn't want to fail, so we can always log the error, and continue.

Reply:
okay makes sense